src/profile-logic/flame-graph.js

```diff
@@ -38,229 +40,255 @@ export type FlameGraphTiming = Array<{
+        const insertionIndex = bisectionLeftByStrKey(
+          rowAtThisDepth,
+          funcName,
+          (cn) => stringTable.getString(funcTableNameColumn[func[cn]]),
+          siblingIndexRangeStart,
+          siblingIndexRangeEnd
+        );
+        rowAtThisDepth.splice(insertionIndex, 0, currentSibling);
+      }
     }
 
-    const funcName = thread.stringTable.getString(
-      thread.funcTable.name[callNodeTable.func[callNodeIndex]]
-    );
+    // Now currentCallNode and all its siblings have been added to the row, and
+    // they are ordered correctly. Find a queued up node in this row which has
+    // children, and descend into it.
+    let candidateDepth = currentDepth;
+    let candidateRow = rowAtThisDepth;
+    let indexInCandidateRow = pendingRangeStartAtDepth[candidateDepth];
+    let candidateNode = candidateRow[indexInCandidateRow];
+    // "while (!hasChildren(candidateNode)) {"
+    while (subtreeRangeEnd[candidateNode] === candidateNode + 1) {
+      // candidateNode does not have any children.
+      // "Finish" candidateNode by incrementing pendingRangeStartAtDepth[candidateDepth].
+      indexInCandidateRow++;
+      pendingRangeStartAtDepth[candidateDepth] = indexInCandidateRow;
 
-    /* From the parent, we can now know the slice allotted for all
-     * its children. */
-    const start = offsets[parent];
-    const end = offsets[parent] + callNodeChildCount[parent] - 1;
-
-    /* Find the place in `array` where this callnode should be
-     * inserted, swapping elements in the array as we go
-     * along. Continue as long as this callnode's function name is
-     * lexically smaller than the function names of the callnodes
-     * already placed in the array. This ensures that all slices have
-     * children in descending order. Any callnode indices equal to 0
-     * means that they are uninitialized, so just breeze through
-     * them. When we stop, when have found the right position to
-     * insert our callnode.
-     *
-     * This effectively is an insertion sort, which is O(n^2), but
-     * since n is typically small (the number of children of a given
-     * callnode), it should be just fine.
-     */
-    let i = start;
-    while (i < end) {
-      if (
-        array[i + 1] !== 0 &&
-        funcName >
-          thread.stringTable.getString(
-            thread.funcTable.name[callNodeTable.func[array[i + 1]]]
-          )
-      ) {
-        // We've found our spot if the next slot in the array is
-        // occupied with a callnode whose function name is less than
-        // ours.
-        break;
+      // Check if the current row contains any other pending call nodes.
+      while (indexInCandidateRow === candidateRow.length) {
+        // There are no more pending nodes in the current row.
+        if (candidateDepth === 0) {
+          // We're completely done.
+          break outer;
+        }
+        // Go up a level and continue the search there.
```

Comment:
do we need to "finish" this level? We do it for the first one above, but what about if we move up several times? It may not be a problem if we never come back at this depth though (it's not clear to me).